src/coins.cpp

```diff
@@ -184,34 +182,33 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
 
 bool CCoinsViewCache::BatchWrite(CoinsViewCacheCursor& cursor, const uint256 &hashBlockIn) {
     for (auto it{cursor.Begin()}; it != cursor.End(); it = cursor.NextAndMaybeErase(*it)) {
-        // Ignore non-dirty entries (optimization).
         if (!it->second.IsDirty()) {
-            continue;
+            throw std::logic_error("A non-DIRTY coin was returned from the cursor in BatchWrite");
+        }
+        if (it->second.IsFresh() && it->second.coin.IsSpent()) {
+            throw std::logic_error("A FRESH coin was not removed when it was spent");
         }
```

Comment:
I find it inconsistent that we're throwing here only for testability - since we assume this cannot happen in reality.
Basically, this is another "return value" of the method, so declaring it a `bool` was essentially a lie (especially since it always seems to just return `true`).
Could we use the return value instead to signal failure?

Reply:
This can happen in our fuzz tests. I think a follow up to this change would be to remove direct access to the flags, and replace with methods `SetDirty` and `SetDirtyAndFresh`. Then we can get rid of fuzzing all kinds of flags and simplify this.

Reply:
Can we consider fixing that before this PR? I wouldn't mind doing parts of that myself!

Reply:
Are you suggesting to refactor direct flags access with `SetDirty` and `SetDirtyAndFresh` first? If we do that first, we would still need a `SetFresh` setter since we still have a case here with setting FRESH-but-not-DIRTY. It's kind of a chicken-and-egg problem.

Reply:
I suggest we make our tests reflect reality before we do this change, since currently if we add asserts about the internal consistency of the code, the tests seem to constantly violate them.